language/stdlib/modules/event.mvir

```diff
@@ -0,0 +1,90 @@
+module Event {
+    import 0x0.Hash;
+    import 0x0.AddressUtil;
+    import 0x0.U64Util;
+    import 0x0.BytearrayUtil;
+
+    // A resource representing the counter used to generate uniqueness under each account
+    resource Counter {
+        // A monotonically increased counter
+        counter: u64,
+    }
+
+    // A handle for a event
+    resource Handle {
+        // Counter for this event stream.
+        counter: u64
+        // A globally unique ID for this event.
+        key: bytearray,
+    }
+
+    get_counter(): &mut R#Self.Counter {
+        let ret: &mut R#Self.Counter;
+        let exists_counter: bool;
+        let sender: address;
+
+        sender = get_txn_sender();
+        exists_counter = exists<Counter>(copy(sender));
+
+        if (!move(exists_counter)) {
+            move_to_sender<Counter>(Counter { count: 0 });
+        }
+        ret = borrow_global<Counter>(move(sender));
+        return move(ret);
+    }
+
+    fresh_key(): bytearray {
+        let counter: &mut R#Self.Counter;
+        let count: &mut u64;
+        let sender: address;
+        let sender_bytes: bytearray;
+        let count_bytes: bytearray;
+        let preimage: bytearray;
+        let ret: bytearray;
+
+        counter = Self.get_counter();
+        count = &mut move(counter).counter;
+        *copy(count) = *copy(count) + 1;
+
+        sender = get_txn_sender();
+        sender_bytes = AddressUtil.address_to_bytes(move(sender));
+        count_bytes = U64Util.u64_to_bytes(*move(count));
+
+        // Counter goes first just in case we want to extend address in the future.
+        preimage = BytearrayUtil.bytearray_concat(move(count_bytes), move(sender_bytes));
+        ret = Hash.sha3_256(move(preimage));
+
+        return move(ret);
+    }
+
+    public new_event_handle(): R#Self.Handle {
+        let fresh_key: bytearray;
+
+        fresh_key = Self.fresh_key();
+        return Handle {key: move(fresh_key), counter: 0};
+    }
+
+    public emit_event(handle_ref: &mut R#Self.Handle, msg: bytearray) {
+        let count: &mut u64;
+        let key: bytearray;
+        key = *&copy(handle_ref).key;
+
+        count = &mut copy(handle_ref).counter;
+        *copy(count) = *copy(count) + 1;
+
+        Self.write_to_event_store(move(key), *move(count) - 1, move(msg));
+        release(move(handle_ref));
+        return;
+    }
+
+    // Native procedure that writes to the actual event stream in Event store
+    // This will replace the "native" portion of EmitEvent bytecode
+    native write_to_event_store(key: bytearray, count: u64, msg: bytearray);
+
+    public destroy(handle: R#Self.Handle) {
+        let key: bytearray;
+        let count: u64;
+        Handle { key, count } = move(handle);
+        return;
+    }
+}
```

Comment:
Do we want that? I feel like Event handles should remain somewhat opaque

Reply:
I suppose you might want to ask something like "did this the counter stored here change since I last checked", which would involve looking at the handle.

My general inclination is to expose getters for everything given that they enable use-cases we can't anticipate. But I could definitely be wrong about this. For example, getters can make it easier to create read/write conflicts that will make it harder for us to parallelize tx execution (if we go that route).